File: 447.number-of-boomerangs.java
/*
 * @lc app=leetcode id=447 lang=java
 *
 * [447] Number of Boomerangs
 */
class Solution {
    public int numberOfBoomerangs(int[][] points) {
        int res = 0;
        if (points == null) return res;
        for (int i = 0; i < points.length; ++i) {
            Map<Integer, Integer> distMap = new HashMap<>();
            for (int j = 0; j < points.length; ++j) {
                if (j == i) continue;
                int key = dist(points[i], points[j]);
                distMap.put(key, distMap.getOrDefault(key, 0) + 1);
            }
            for (int count : distMap.values()) {
                res += (count - 1) * count;
            }
        }
        return res;
    }

    private int dist(int[] p1, int[] p2) {
        return (p1[0]-p2[0])*(p1[0]-p2[0])+(p1[1]-p2[1])*(p1[1]-p2[1]);
    }
}

File: go/447.number-of-boomerangs.go
/*
 * @lc app=leetcode id=447 lang=golang
 *
 * [447] Number of Boomerangs
 */

// @lc code=start
func numberOfBoomerangs(points [][]int) int {
	res := 0
	for i := 0; i < len(points); i++ {
		cnt := make(map[int]int)
		for j := 0; j < len(points); j++ {
			if j == i {
				continue
			}
			cnt[d(points[i], points[j])]++
		}
		for _, c := range cnt {
			res += c * (c - 1)
		}
	}
	return res
}

func d(p, q []int) int {
	i, j := p[0]-q[0], p[1]-q[1]
	return i*i + j*j
}

// @lc code=end

